crates/noirc_evaluator/src/ssa_refactor/ssa_gen/context.rs

```diff
@@ -255,6 +259,88 @@ impl<'a> FunctionContext<'a> {
         result.into()
     }
 
+    /// The frontend claims to support equality (==) on arrays, so we must support it in SSA here.
+    /// The actual BinaryOp::Eq in SSA is meant only for primitive numeric types so we encode an
+    /// entire equality loop on each array element. The generated IR is as follows:
+    ///
+    ///   ...
+    ///   result_alloc = allocate
+    ///   store u1 1 in result_alloc
+    ///   jmp loop_start(0)
+    /// loop_start(i: Field):
+    ///   v0 = lt i, array_len
+    ///   jmpif v0, then: loop_body, else: loop_end
+    /// loop_body():
+    ///   v1 = array_get lhs, index i
+    ///   v2 = array_get rhs, index i
+    ///   v3 = eq v1, v2
+    ///   v4 = load result_alloc
+    ///   v5 = and v4, v3
+    ///   store v5 in result_alloc
+    ///   v6 = add i, Field 1
+    ///   jmp loop_start(v6)
+    /// loop_end():
+    ///   result = load result_alloc
+    fn insert_array_equality(
+        &mut self,
+        lhs: ValueId,
+        operator: noirc_frontend::BinaryOpKind,
+        rhs: ValueId,
+    ) -> Values {
+        let lhs_type = self.builder.type_of_value(lhs);
+        let rhs_type = self.builder.type_of_value(rhs);
+
+        let (array_length, element_type) = match (lhs_type, rhs_type) {
+            (Type::Array(lhs_elements, lhs_length), Type::Array(rhs_elements, rhs_length)) => {
+                assert_eq!(lhs_elements.len(), 1, "== is unimplemented for arrays of structs");
+                assert_eq!(rhs_elements.len(), 1, "== is unimplemented for arrays of structs");
+                assert_eq!(lhs_elements[0], rhs_elements[0]);
```

Comment:
Instead of `lhs_elements` & `rhs_elements`, could we call them `lhs_composite_type` & `lhs_composite_type`. Otherwise the length == 1 assert is unintuitive to read.

Reply:
Hope you'll forgive me for editing - wanted to get this merged!